diffsims/pattern/detector_functions.py

```diff
@@ -0,0 +1,241 @@
+    """
+    Adds shot noise (optional) and gaussian point spread (via a convolution) to a pattern
+
+    Parameters
+    ----------
+    pattern : np.array
+        The diffraction pattern at the detector
+    sigma : float
+        The standard deviation of the gaussian blur, in pixels
+    shot_noise : bool
+        Whether to include shot noise in the original signal, default True
+    seed : int or None
+        seed value for the random number generator (effects the shot noise only)
+
+    Returns
+    -------
+    detector_pattern : np.array
+        A single sample of the pattern after accounting for detector properties
+
+    See also
+    --------
+    add_shot_noise : adds only shot noise
+    add_gaussian_point_spread
+    """
+
+    # shot noise happens before the detector response (operations won't commute)
+    if shot_noise:
+        pattern = add_shot_noise(pattern, seed)
+
+    pattern = add_gaussian_point_spread(pattern, sigma)
+
+    return pattern
+
+
+def add_gaussian_noise(pattern, sigma, seed=None):
+    """
+    Applies gaussian noise at each pixel within the pattern
+
+    Parameters
+    ----------
+    pattern : np.array
+        The diffraction pattern at the detector
+    sigma : float
+        The (absolute) deviation of the gaussian errors
+    seed : int or None
+        seed value for the random number generator
+
+    Returns
+    -------
+    corrupted_pattern :
+    """
+    rng = _process_seed_argument(seed)
+    pertubations = rng.normal(loc=0, scale=sigma, size=pattern.shape)
+    pattern = pattern + pertubations
+
+    return constrain_to_dynamic_range(pattern)
+
+
+def add_dead_pixels(pattern, n=None, fraction=None, seed=None):
+    """
+    Adds randomly placed dead pixels onto a pattern
+
+    Parameters
+    ----------
+    pattern : np.array
+        The diffraction pattern at the detector
+    n : int
+        The number of dead pixels, defaults to None
+    fraction : float
+        The fraction of dead pixels, defaults to None
+    seed : int or None
+        seed value for the random number generator
+
+    Returns
+    -------
+    corrupted_pattern : np.array
+        The pattern, with dead pixels included
+    """
+    # sorting the n/fraction kwargs
+
+    both_none = n is None and fraction is None
+    neither_none = n is not None and fraction is not None
+
+    if both_none or neither_none:
+        raise ValueError("Exactly one of 'n' and 'fraction' must be None")
+
+    # converting fraction to n if needs be
+    if fraction is not None:
+        pattern_size = pattern.shape[0] * pattern.shape[1]
+        n = int(fraction * pattern_size)
+
+    rng = _process_seed_argument(seed)
+
+    # .astype rounds down, these generate values from 0 to (pattern.shape - 1)
+    xdead = rng.uniform(low=0, high=pattern.shape[0], size=n).astype(int)
+    ydead = rng.uniform(low=0, high=pattern.shape[1], size=n).astype(int)
+
+    pattern[xdead, ydead] = 0
```

Comment:
This shouldn't operate inplace, but on new copy of the pattern (deepcopy()).

I think this should be the case for all functions.

Reply:
I think this bug only happens in this function, but I will write tests for all of them and fix this

Reply:
np.may_share_memory(array1, array2) returning true/false is a test i use often.